### PR TITLE
feat!(node): add support for Node18, drop Node12, update Node16

### LINKED ISSUE #2255: Update supported Node versions
**Describe the enhancement**
Currently, JavaScript composite actions only support Node 12 and 16.
Node 12 is deprecated and Node 16 has an accelerated EoL. So arguably, none are viable choices for long-term development.

I would propose removing support for Node 12 and adding support for 18 (current Active LTS).

### PR DESCRIPTION
**Pull request by louis-bompart**
Fixes #2255 

## Changes
- **BREAKING:** Remove support for Node 12 (and all the code relative to it)
- Add support for Node 18
- Updated Node 16 version

## Notes:
 - Uses https://github.com/louis-bompart/runner_L0/ for the fixtures. Some changes were required (updates of dependencies+ node bump 12 to 18).
  I recommend moving the fixtures in this repo: this will allow public contributors to have a one-stop shop (or just to clone another branch, and not a whole other repo).
 - Didn't manually test. Relied 💯 on CI. See [run](https://github.com/louis-bompart/runner/actions/runs/3467460136/jobs/5792350319)

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@ on:
 jobs:
   build:
     strategy:
+      fail-fast: false
```
Comment (louis-bompart) [inline .github/workflows/build.yml:20]: derp
```suggestion
```

File: src/Misc/externals.sh [hunk 2]
```diff
@@ -4,8 +4,10 @@ PRECACHE=$2
 
 NODE_URL=https://nodejs.org/dist
 UNOFFICIAL_NODE_URL=https://unofficial-builds.nodejs.org/download/release
-NODE12_VERSION="12.22.7"
-NODE16_VERSION="16.13.0"
+# https://github.com/nodejs/build/issues/2540
```
Comment (louis-bompart) [inline src/Misc/externals.sh:7]: To monitor: https://github.com/nodejs/build/issues/3077
Win-arm64 official build should be discussed in this meeting/issue

File: src/Misc/externals.sh [hunk 3]
```diff
@@ -147,29 +149,29 @@ fi
...
     acquireExternalTool "$NODE_URL/v${NODE16_VERSION}/node-v${NODE16_VERSION}-darwin-arm64.tar.gz" node16 fix_nested_dir
+    acquireExternalTool "$NODE_URL/v${NODE18_VERSION}/node-v${NODE18_VERSION}-darwin-arm64.tar.gz" node18 fix_nested_dir
 fi
 
 # Download the external tools for Linux PACKAGERUNTIMEs.
 if [[ "$PACKAGERUNTIME" == "linux-x64" ]]; then
-    acquireExternalTool "$NODE_URL/v${NODE12_VERSION}/node-v${NODE12_VERSION}-linux-x64.tar.gz" node12 fix_nested_dir
-    acquireExternalTool "https://vstsagenttools.blob.core.windows.net/tools/nodejs/${NODE12_VERSION}/alpine/x64/node-v${NODE12_VERSION}-alpine-x64.tar.gz" node12_alpine
     acquireExternalTool "$NODE_URL/v${NODE16_VERSION}/node-v${NODE16_VERSION}-linux-x64.tar.gz" node16 fix_nested_dir
-    acquireExternalTool "https://vstsagenttools.blob.core.windows.net/tools/nodejs/${NODE16_VERSION}/alpine/x64/node-v${NODE16_VERSION}-alpine-x64.tar.gz" node16_alpine
```
Comment (louis-bompart) [inline src/Misc/externals.sh:164]: Yeah so. I didn't find the version I needed in vstsagenttools.blob.core.windows.net
So I just replaced it with the good ol' build from bona fide NodeJS.

File: src/Misc/externals.sh (hunk 2 above)
Comment (sxa) [reply src/Misc/externals.sh:7]: FYI We're working towards it in https://github.com/nodejs/build/issues/3046#issuecomment-1315211429 but as always with a new platform it's dependent on people and systems to be able to elevate it to an official release, so for now pulling from the unofficial builds repo is the only way to get a versioned download from the nodejs project.

Comment (louis-bompart) [reply src/Misc/externals.sh:7]: Thanks for the heads-up @sxa , I somehow missed this one while digging up 😅

**Comment by remal** [comment]
@fhammerl, what's the plan for having Node 18? Can we have node18 runtime for GitHub actions?

**Comment by fhammerl** [comment]
@remal With Node16 security updates coming to an end this October, we're of course planning on embedding a new Node.js version, but I can't give an exact date as of today.

**Comment by MikeMcC399** [comment]
@fhammerl 

> With Node16 security updates coming to an end this October, we're of course planning on embedding a new Node.js version, but I can't give an exact date as of today.

Perhaps you could answer into https://github.com/orgs/community/discussions/53217 when your plans are more concrete? Especially as this here is a closed PR and enhancement requests are specifically pointed to the discussion groups according to the [issues template](https://github.com/actions/runner/blob/main/.github/ISSUE_TEMPLATE/config.yml).

[Node.js Release Schedule](https://github.com/nodejs/release#release-schedule) states Sep 11, 2023 as end-of-life transition for Node.js `16`, which is earlier than the October date which you mentioned.

**Comment by TomasHubelbauer** [comment]
I'd be very grateful for more engagement in open community discussions as at the moment for people interested in this update, there are two closed issues with maintainer activity one has to discover and subscribe to themselves (this and #2255) whereas the open and officially recommended discussion forum sees no maintainer activity and information shared there is dug for these closed threads.

It works when one puts effort into discovering this information but it feels like things should flow the opposite direction. Just a friendly request though, at the end of the day I am happy with any information we get regardless where it comes from 🙏

**Pull request closed without merging**